scripts/sensor/_04b_time_frequency_csp.py

```diff
@@ -0,0 +1,436 @@
+    y = _prepare_labels(epochs=epochs_filt, contrast=contrast)
+
+    return epochs_filt, y
+
+
+@failsafe_run(on_error=on_error, script_path=__file__)
+def one_subject_decoding(
+    *,
+    cfg,
+    subject: str,
+    session: str,
+    contrast: Tuple[str, str],
+
+) -> None:
+    """Run one subject.
+
+    There are two steps in this function:
+    1. The frequency analysis.
+    2. The time-frequency analysis.
+    """
+    condition1, condition2 = contrast
+    msg = f'Contrasting conditions: {condition1} – {condition2}'
+    logger.info(**gen_log_kwargs(msg, subject=subject, session=session))
+
+    # XXX The following 30 or so lines are actually identical to code we have
+    # in _03_decoding_time_by_time. We should try to refactor this to reduce
+    # duplication.
+    fname_epochs = BIDSPath(subject=subject,
+                            session=session,
+                            task=cfg.task,
+                            acquisition=cfg.acq,
+                            run=None,
+                            recording=cfg.rec,
+                            space=cfg.space,
+                            suffix='epo',
+                            extension='.fif',
+                            datatype=cfg.datatype,
+                            root=cfg.deriv_root,
+                            check=False)
+
+    epochs = mne.read_epochs(fname_epochs)
+    if cfg.analyze_channels:
+        # We special-case the average reference here to work around a situation
+        # where e.g. `analyze_channels` might contain only a single channel:
+        # `concatenate_epochs` below will then fail when trying to create /
+        # apply the projection. We can avoid this by removing an existing
+        # average reference projection here, and applying the average reference
+        # directly – without going through a projector.
+        if 'eeg' in cfg.ch_types and cfg.eeg_reference == 'average':
+            epochs.set_eeg_reference('average')
+        else:
+            epochs.apply_proj()
+        epochs.pick(cfg.analyze_channels)
+
+    if cfg.time_frequency_subtract_evoked:
+        epochs.subtract_evoked()
+
+    # Perform rank reduction via PCA.
+    #
+    # Select the channel type with the smallest rank.
+    # Limit it to a maximum of 100.
+    ranks = mne.compute_rank(inst=epochs, rank='info')
+    ch_type_smallest_rank = min(ranks, key=ranks.get)
+    rank = min(
+        ranks[ch_type_smallest_rank],
+        100
+    )
+    del ch_type_smallest_rank, ranks
+
+    msg = f'Reducing data dimension via PCA; new rank: {rank}.'
+    logger.info(**gen_log_kwargs(msg, subject=subject, session=session))
+    pca = UnsupervisedSpatialFilter(
+        PCA(rank),
+        average=False
+    )
+
+    # Classifier
+    csp = CSP(
+        n_components=4,  # XXX revisit
+        reg=0.1,         # XXX revisit
+        rank='info',
+    )
+    clf = make_pipeline(
+        csp,
+        LinearDiscriminantAnalysis(),
```

Comment:
I would say we default to LogReg like in other decoding. Unless there is strong evidence that it's worse

Reply:
Done, let's see how that flies…